lookout/style/format/benchmarks/quality_report.py

```diff
@@ -430,44 +429,41 @@ def generate_quality_report(input: str, output: str, force: bool, bblfsh: str, t
-                        table[1:], col=fields2id[field], weight_col=fields2id[weight_field]))
-                table.append(("average",) + average)
-                table.append(("weighted average",) + tuple(weighted_average))
-                float_fields = ("precision", "recall", "full_recall", "f1", "full_f1", "ppcr")
-                floatfmts = []
-                for field in fields2id:
-                    if field in float_fields:
-                        floatfmts.append(FLOAT_PRECISION)
-                    elif field == "Average Rule Len":
-                        floatfmts.append(".1f")
-                    else:
-                        floatfmts.append("g")
+            for repo, report in reports:
+                metrics = _get_metrics(getattr(report, report_name))
+                if not table:
+                    table.append(("repo",) + metrics._fields + additional_fields)
+                    for i, field in enumerate(table[0]):
+                        fields2id[field] = i
+                n_rules, avg_len = _get_model_summary(report.model_report)
+                table.append((get_repo_name(repo),) + metrics + (n_rules, avg_len))
+            avgvals = tuple(calc_avg(table[1:], fields2id[field]) for field in metrics._fields)
+            average = tuple(("%" + FLOAT_PRECISION) % v for v in avgvals[:-2])
+            average += tuple("%d" % v for v in avgvals[-2:])  # support, full_support
+            average += tuple(("%d", "%.1f")[i] % calc_avg(table[1:], fields2id[field])
+                             for i, field in enumerate(additional_fields))
+            fields_to_weight = (
+                ("precision", "support"), ("recall", "support"),
+                ("full_recall", "full_support"), ("f1", "support"),
+                ("full_f1", "full_support"), ("ppcr", "support"),
+            )
+            weighted_average = []
+            for field, weight_field in fields_to_weight:
+                weighted_average.append(("%" + FLOAT_PRECISION) % calc_weighted_avg(
+                    table[1:], col=fields2id[field], weight_col=fields2id[weight_field]))
+            table.append(("average",) + average)
+            table.append(("weighted average",) + tuple(weighted_average))
+            float_fields = ("precision", "recall", "full_recall", "f1", "full_f1", "ppcr")
+            floatfmts = []
+            for field in fields2id:
+                if field in float_fields:
+                    floatfmts.append(FLOAT_PRECISION)
+                elif field == "Average Rule Len":
+                    floatfmts.append(".1f")
+                else:
+                    floatfmts.append("g")
 
-                print(tabulate(table, tablefmt="pipe", headers="firstrow", floatfmt=floatfmts),
-                      file=output)
-                summary = output.getvalue()
+            summary = tabulate(table, tablefmt="pipe", headers="firstrow", floatfmt=floatfmts)
```

Comment:
May I ask you to improve this even further: return the string table from the function.

Then in the higher level code:

```python
summary = generate_quality_report(...)
log.info("%s", report_name)
log.info("\n%s", summary)
summary_loc = os.path.join(output, "summary-%s.md" % report_name)
with open(summary_loc, "w", encoding="utf-8") as f:
    f.write(summary)
```

We can put it to another function, say `write_quality_report`

Reply:
good point.